### PR TITLE
wgsl abs() built-in function execution test

### PR DESCRIPTION
**Pull request by sarahM0**
<hr>

**Author checklist for test code/plans:**

- [ ] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [ ] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [ ] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [x] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [x] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [x] Existing (or new) test helpers are used where they would reduce complexity.
- [x] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/shader/util.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+export const PositiveInf = 0x7f800000;
+export const NegativeInf = 0xff800000;
```
Comment (kainino0x) [inline src/webgpu/shader/util.ts:2]: nit: kConstantName

nit: to be more descriptive, kPositiveInfF32AsInteger, or something like that?

nit: src/webgpu/util/math.ts would probably be a good place for these

File: src/webgpu/shader/util.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  const i32Arr = new Int32Array(f32Arr.buffer);
+  return i32Arr[0];
+}
+
+export function uint32ToInt32(u32: number): number {
+  const u32Arr = new Uint32Array(1);
+  u32Arr[0] = u32;
+  const i32Arr = new Int32Array(u32Arr.buffer);
+  return i32Arr[0];
+}
```
Comment (kainino0x) [inline src/webgpu/shader/util.ts:42]: I think `src/webgpu/util/conversion.ts` would be a good place for these helpers. I don't think it duplicates any of them, they're all more complicated.

File: src/webgpu/shader/util.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  Float,
+  Int,
+  Uint,
+  Hex,
+}
+
+export type Case = {
+  input: number;
+  expected: number[];
+};
```
Comment (kainino0x) [inline src/webgpu/shader/util.ts:14]: This seems specific to builtin.ts?

File: src/webgpu/shader/execution/builtin/builtin.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      break;
+    }
+    case NumberType.Hex: {
+      inputData = new Uint32Array(4 * cases.length);
+      break;
+    }
+    case NumberType.Uint: {
+      inputData = new Uint32Array(4 * cases.length);
+      break;
+    }
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/builtin.ts:36]: nit:
```suggestion
    case NumberType.Hex:
    case NumberType.Uint: {
      inputData = new Uint32Array(4 * cases.length);
      break;
    }
```

File: src/webgpu/shader/execution/builtin/builtin.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+//   switch(numberType){{
+//     case Float:
+//       const data : Data{asF32 = inData, asU32= }
+//       return DataTransfer
+//   }
+//   const expectedDataAsU32 = numberType === NumberType.Float ? float32ToUint32(e) : e;
+//   const expectedDataAsF32 = numberType === NumberType.Float ? e : uint32ToFloat32(e);
+//   const expectedDataAsI32 =
+//     numberType === NumberType.Float ? float32ToInt32(e) : uint32ToInt32(e);
+// }
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/builtin.ts:104]: nit: commented out code

**Review by kainino0x — COMMENTED** [review]
Had assorted comments, but haven't actually had a chance to review yet, sorry

**Comment by sarahM0** [comment]
Please review [this](https://github.com/gpuweb/cts/pull/731) first. When that merges I'll rebase this PR and it will be ready for review. Sorry for the confusion.

File: src/webgpu/shader/execution/builtin/builtin.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+export type Case = {
+  input: number;
+  expected: number[];
+};
+
+export const g = makeTestGroup(GPUTest);
+
+type Cases = Case[];
+
+export function createInputBuffer(
```
Comment (sarahM0) [inline src/webgpu/shader/execution/builtin/builtin.ts:28]: I'll improve this function in my next PR.

File: src/webgpu/shader/execution/builtin/abs.spec.ts [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+`
+  )
+  .params(u =>
+    u
+      .combineWithParams([
+        { storageClass: 'storage', storageMode: 'read_write', access: 'read' },
+      ] as const)
+      .combineWithParams([{ containerType: 'scalar' }, { containerType: 'vector' }] as const)
+      .combineWithParams([{ isAtomic: false }] as const)
+      .combineWithParams([{ baseType: 'f32' }] as const)
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/abs.spec.ts:33]: nit:
```suggestion
      .combine('containerType', ['scalar', 'vector'] as const)
      .combine('isAtomic', [false])
      .combine('baseType', ['f32'] as const)
```
(unless there are plans for expansion that would have to change these back to combineWithParams

File: src/webgpu/shader/execution/builtin/abs.spec.ts [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        { input: -131072, expected: [131072] },
+        { input: -262144, expected: [262144] },
+        { input: -524288, expected: [524288] },
+        { input: -1048576, expected: [1048576] },
+        { input: -8388607, expected: [8388607] },
+        { input: -16777215, expected: [16777215] },
+        { input: -16777216, expected: [16777216] },
+        { input: -134217727, expected: [134217727] },
+        { input: -1073741823, expected: [1073741823] },
+        { input: -2147483647, expected: [2147483647] },
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/abs.spec.ts:163]: Does `abs(-2147483648)` have a defined value that we can test? (I would guess it's `-2147483648`?)

File: src/webgpu/shader/execution/builtin/abs.spec.ts [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        { input: 65536, expected: [65536] },
+        { input: 131072, expected: [131072] },
+        { input: 262144, expected: [262144] },
+        { input: 524288, expected: [524288] },
+        { input: 1048576, expected: [1048576] },
+        { input: 8388607, expected: [8388607] }, //2^23 - 1
+        { input: 16777215, expected: [16777215] }, //2^24 - 1
+        { input: 16777216, expected: [16777216] }, //2^24
+        { input: 134217727, expected: [134217727] }, //2^27 - 1
+        { input: 1073741823, expected: [1073741823] }, //2^30 - 1
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/abs.spec.ts:190]: nit: Case for 2147483647?

also nit: Can write cases as expressions or `0x` literals as it might be clearer. E.g. `0xFFFF_FFFF` or `-0x1_0000_0000`

File: src/webgpu/shader/execution/builtin/builtin.ts [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+  const errs: string[] = [];
+  for (let i = 0; i < cases.length; i++) {
+    const input: string[] = [];
+    const output: string[] = [];
+    const expected: string[] = [];
+    let matched = true;
+    for (let j = 0; j < arrayLength; j++) {
+      const idx = i * 4 + j;
+      const caseExpected: string[] = [];
+      cases[i].expected.forEach((e: number) => {
```
Comment (kainino0x) [inline src/webgpu/shader/execution/builtin/builtin.ts:130]: nit: prefer a for-each loop: `for (const e of cases[i]) {`

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/shader/execution/builtin/builtin.ts (hunk 10 above)
Comment (sarahM0) [reply src/webgpu/shader/execution/builtin/builtin.ts]: done

File: src/webgpu/shader/execution/builtin/abs.spec.ts (hunk 9 above)
Comment (sarahM0) [reply src/webgpu/shader/execution/builtin/abs.spec.ts:190]: next cl

File: src/webgpu/shader/execution/builtin/abs.spec.ts (hunk 8 above)
Comment (sarahM0) [reply src/webgpu/shader/execution/builtin/abs.spec.ts:163]: defined value? you mean define a constant and use that instead of the gibberish number? (if yes, then my next pr fixes it).

Comment (kainino0x) [reply src/webgpu/shader/execution/builtin/abs.spec.ts:163]: I just mean does WGSL's definition of abs() have a well-specified result for `abs(-0x8000_0000)`.
And in fact it does: https://gpuweb.github.io/gpuweb/wgsl/#integer-builtin-functions
> If e evaluates to the largest negative value, then the result is e.

Comment (kainino0x) [reply src/webgpu/shader/execution/builtin/abs.spec.ts:163]: So we should test that corner case.

**Review by kainino0x — APPROVED** [review]
